File: Ruby/Work/employee.rb
#!/usr/bin/ruby

#####################
#
# 	Created by: breton 
#	File: employee.rb
#
#####################

class Employee
    def initialize(name, payrate)
        @name = name
        @payrate = payrate
    end
    
    def name
        return @name
    end

    def pay(hours)
        return @payrate * hours
    end
end

class Manager < Employee
    def initialize(name, payrate, salary)
        super(name, payrate)
        @salary = salary
    end
    
    def pay(hours)
        if @salary
            return @payrate
            
        else
            return @payrate *hours
        end
    end
            
    
end

e1 = Employee.new("Brown", 20)
print(e1.name, " pay: ", e1.pay(40), "\n")

m1 = Manager.new("Smith", 1200, true)
print(m1.name, " Pay: ", m1.pay(0), "\n")
m2 = Manager.new("Jones",40,false)
print(m2.name, " pay: ", m2.pay(40))

File: Python/ObjectOP/employee.py
#!/usr/bin/python

# Base class
class Employee:
    def  __init__(self, name, payRate):
        self.name = name
        self.payRate = payRate

    def __str__(self):
        return self.name + ", " + str(self.payRate)

    def pay(self, hoursWorked):
        return self.payRate * hoursWorked

# Drive class - inherites fields and methods from base class
class Manager(Employee):
    def __init__(self, name, payRate, isSalaried): #constructor
        Employee.__init__(self, name, payRate) #call base class constructor with arguments needed
        self.salaried = isSalaried #field set to isSalaried paramater in constructor

    #2-string method
    def __str__(self):
        retStr = Employee.__str__(self)
        retStr += " salaried: " + str(self.salaried)
        return retStr

    def pay(self, hoursWorked):
        if self.salaried:
            return self.payRate
        else:
            return self.payRate * hoursWorked


e1 = Employee("John Jones", 10.00)
print(e1)
print("Gross pay: " + str(e1.pay(40)))

m1 = Manager("Jane Smith", 1200, True) #this is a salaried manager
print(m1)
print("Gross: " + str(m1.pay(40)))

m2 = Manager("Jim Brown", 20.00, False) #not a salaried manager
print(m2)
print("Gross: " + str(m2.pay(40)))
